File: app/src/androidTest/java/pavelnazimok/uitestingfeatures/java/utils/EspressoExtensions.java
package pavelnazimok.uitestingfeatures.java.utils;

import android.os.IBinder;
import android.view.View;
import android.view.WindowManager;
import android.widget.Checkable;
import android.widget.ListView;
import android.widget.TextView;

import androidx.test.espresso.NoMatchingViewException;
import androidx.test.espresso.Root;
import androidx.test.espresso.UiController;
import androidx.test.espresso.ViewAction;
import androidx.test.espresso.matcher.BoundedMatcher;
import androidx.test.espresso.matcher.ViewMatchers;

import org.hamcrest.BaseMatcher;
import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import java.util.regex.Pattern;

import static androidx.test.espresso.Espresso.onData;
import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isAssignableFrom;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withEffectiveVisibility;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static com.schibsted.spain.barista.interaction.BaristaSleepInteractions.sleepThread;
import static com.schibsted.spain.barista.internal.util.ResourceTypeKt.resourceMatcher;
import static org.assertj.core.api.Assertions.assertThat;
import static org.hamcrest.CoreMatchers.allOf;
import static org.hamcrest.CoreMatchers.anything;
import static org.hamcrest.CoreMatchers.isA;
import static org.hamcrest.CoreMatchers.not;
import static pavelnazimok.uitestingfeatures.java.utils.AndroidExtensions.resId;
import static pavelnazimok.uitestingfeatures.java.utils.AndroidExtensions.resText;

public class EspressoExtensions {

    public static void waitForView(Matcher<View> matcher, long timeout) {
        long endTime = System.currentTimeMillis() + timeout;

        while (System.currentTimeMillis() < endTime) {
            try {
                onView(matcher).check(matches(isDisplayed()));
                return;
            } catch (NoMatchingViewException e) {
                sleepThread(500);
            }
        }

        throw new EspressoTimeoutException(String.format(
                "The View '%s' was not found within %d millis",
                matcher.toString(),
                timeout
        ));
    }

    public static void waitForView(int resId, long timeout) {
        waitForView(resourceMatcher(resId), timeout);
    }

    public static void waitForView(String text, long timeout) {
        waitForView(withText(text), timeout);
    }

    public static void waitForView(int viewId, int stringId, long timeout) {
        waitForView(allOf(withId(viewId), withText(stringId)), timeout);
    }

    public static void waitForView(int viewId, String text, long timeout) {
        waitForView(allOf(withId(viewId), withText(text)), timeout);
    }

    public static void waitForViewWithPattern(Matcher<View> matcher, Pattern pattern, long timeout) {
        long endTime = System.currentTimeMillis() + timeout;

        while (System.currentTimeMillis() < endTime) {
            try {
                onView(matcher).check(matches(withPattern(pattern)));
                return;
            } catch (NoMatchingViewException | AssertionError e) {
                sleepThread(500);
            }
        }

        onView(matcher).check(matches(withPattern(pattern)));
    }

    public static void waitForViewWithPattern(int viewId, Pattern pattern, long timeout) {
        waitForViewWithPattern(withId(viewId), pattern, timeout);
    }

    public static Matcher<View> withPattern(Pattern pattern) {
        return new BoundedMatcher<View, TextView>(TextView.class) {
            @Override
            public void describeTo(Description description) {
                description
                        .appendText("with regex pattern: ")
                        .appendValue(pattern.toString());
            }

            @Override
            protected boolean matchesSafely(TextView item) {
                return pattern.matcher(item.getText()).find();
            }
        };
    }

    public static void waitForTextInView(String text, int viewId, long timeout) {
        long endTime = System.currentTimeMillis() + timeout;
        String textInView = "";

        while (System.currentTimeMillis() < endTime) {
            textInView = getText(viewId);
            if (textInView.equals(text)) break;
            sleepThread(500);
        }

        assertThat(textInView)
                .as(String.format("Expected text was not found in the View with ID '%s' within %d millis", resId(viewId), timeout))
                .isEqualTo(text);
    }

    public static void waitForTextInView(int stringId, int viewId, long timeout) {
        waitForTextInView(resText(stringId), viewId, timeout);
    }

    public static void waitForToast(String text, long timeout) {
        Matcher<Root> matcher = new TypeSafeMatcher<Root>() {
            @Override
            public void describeTo(Description description) {
                description.appendText("is Toast");
            }

            @Override
            protected boolean matchesSafely(Root item) {
                int type = item.getWindowLayoutParams().get().type;
                if (type == WindowManager.LayoutParams.TYPE_TOAST) {
                    IBinder windowToken = item.getDecorView().getWindowToken();
                    IBinder appToken = item.getDecorView().getApplicationWindowToken();
                    return (windowToken == appToken);
                }
                return false;
            }
        };

        long endTime = System.currentTimeMillis() + timeout;
        while (System.currentTimeMillis() < endTime) {
            try {
                onView(withText(text)).inRoot(matcher).check(matches(isDisplayed()));
                return;
            } catch (NoMatchingViewException e) {
                sleepThread(500);
            }
        }

        throw new EspressoTimeoutException(String.format("The Toast with text '%s' was not found within %d millis", text, timeout));
    }

    public static void setChecked(Matcher<View> matcher, boolean check) {
        onView(matcher).perform(new ViewAction() {
            @Override
            public BaseMatcher<View> getConstraints() {
                return new BaseMatcher<View>() {
                    @Override
                    public boolean matches(Object item) {
                        return isA(Checkable.class).matches(item);
                    }

                    @Override
                    public void describeMismatch(Object item, Description mismatchDescription) {}

                    @Override
                    public void describeTo(Description description) {}
                };
            }

            @Override
            public String getDescription() {
                return null;
            }

            @Override
            public void perform(UiController uiController, View view) {
                Checkable checkableView = (Checkable) view;

                if (check && !checkableView.isChecked()) {
                    view.performClick();
                } else if (!check && checkableView.isChecked()) {
                    view.performClick();
                }
            }
        });
    }

    public static void setChecked(int viewId, boolean check) {
        setChecked(withId(viewId), check);
    }

    public static String getText(Matcher<View> matcher) {
        AtomicReference<String> text = new AtomicReference<>();

        onView(allOf(matcher, not(withEffectiveVisibility(ViewMatchers.Visibility.GONE)))).perform(new ViewAction() {
            @Override
            public Matcher<View> getConstraints() {
                return isAssignableFrom(TextView.class);
            }

            @Override
            public String getDescription() {
                return "getting text from a TextView";
            }

            @Override
            public void perform(UiController uiController, View view) {
                text.set(((TextView) view).getText().toString());
            }
        });

        return text.get();
    }

    public static String getText(int viewId) {
        return getText(withId(viewId));
    }

    public static int getListViewItemsCount(int listViewId) {
        AtomicInteger count = new AtomicInteger(0);

        Matcher<View> matcher = new TypeSafeMatcher<View>() {
            @Override
            public void describeTo(Description description) {
            }

            @Override
            protected boolean matchesSafely(View item) {
                count.set(((ListView) item).getCount());
                return true;
            }
        };
        onView(withId(listViewId)).check(matches(matcher));

        return count.get();
    }

    public static String getListViewItemText(int listViewId, int position) {
        AtomicReference<String> text = new AtomicReference<>();

        onData(anything())
                .inAdapterView(withId(listViewId))
                .atPosition(position)
                .perform(new ViewAction() {
                    @Override
                    public Matcher<View> getConstraints() {
                        return isAssignableFrom(TextView.class);
                    }

                    @Override
                    public String getDescription() {
                        return "getting text from a ListView item";
                    }

                    @Override
                    public void perform(UiController uiController, View view) {
                        text.set(((TextView) view).getText().toString());
                    }
                });

        return text.get();
    }
}

File: app/src/androidTest/java/pavelnazimok/uitestingfeatures/kotlin/utils/EspressoExtensions.kt
package pavelnazimok.uitestingfeatures.kotlin.utils

import android.view.View
import android.view.WindowManager
import android.widget.Checkable
import android.widget.ListView
import android.widget.TextView
import androidx.annotation.IdRes
import androidx.annotation.StringRes
import androidx.test.espresso.Espresso.onData
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.NoMatchingViewException
import androidx.test.espresso.Root
import androidx.test.espresso.UiController
import androidx.test.espresso.ViewAction
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.BoundedMatcher
import androidx.test.espresso.matcher.ViewMatchers.isAssignableFrom
import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import com.schibsted.spain.barista.interaction.BaristaSleepInteractions.sleepThread
import com.schibsted.spain.barista.internal.util.resourceMatcher
import junit.framework.AssertionFailedError
import org.assertj.core.api.Assertions.assertThat
import org.hamcrest.BaseMatcher
import org.hamcrest.CoreMatchers
import org.hamcrest.CoreMatchers.anything
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.Matchers.allOf
import org.hamcrest.TypeSafeMatcher
import java.util.concurrent.TimeoutException
import java.util.regex.Pattern

fun waitForView(matcher: Matcher<View>, timeout: Long = 20000) {
    val endTime = System.currentTimeMillis() + timeout

    while (System.currentTimeMillis() < endTime) {
        try {
            onView(matcher).check(matches(isDisplayed()))
            return
        } catch (e: NoMatchingViewException) {
            sleepThread(500)
        } catch (e: AssertionFailedError) {
            sleepThread(500)
        }
    }

    throw TimeoutException("The View $matcher was not found within $timeout millis")
}

fun waitForView(resId: Int, timeout: Long = 20000) {
    waitForView(resId.resourceMatcher(), timeout)
}

fun waitForView(text: String, timeout: Long = 20000) {
    waitForView(withText(text), timeout)
}

fun waitForView(@IdRes viewId: Int, @StringRes stringId: Int, timeout: Long = 20000) {
    waitForView(allOf(withId(viewId), withText(stringId)), timeout)
}

fun waitForView(@IdRes viewId: Int, text: String, timeout: Long = 20000) {
    waitForView(allOf(withId(viewId), withText(text)), timeout)
}

fun waitForViewWithPattern(matcher: Matcher<View>, pattern: Pattern, timeout: Long) {
    val endTime = System.currentTimeMillis() + timeout

    while (System.currentTimeMillis() < endTime) {
        try {
            onView(matcher).check(matches(withPattern(pattern)))
            return
        } catch (e: NoMatchingViewException) {
            sleepThread(500)
        } catch (e: AssertionError) {
            sleepThread(500)
        }
    }

    onView(matcher).check(matches(withPattern(pattern)))
}

fun waitForViewWithPattern(@IdRes viewId: Int, pattern: Pattern, timeout: Long) {
    waitForViewWithPattern(withId(viewId), pattern, timeout)
}

fun withPattern(pattern: Pattern): Matcher<View> {
    return object : BoundedMatcher<View, TextView>(TextView::class.java) {
        override fun describeTo(description: Description) {
            description
                    .appendText("with regex pattern: ")
                    .appendValue(pattern.toString())
        }

        override fun matchesSafely(item: TextView): Boolean {
            return pattern.matcher(item.text).find()
        }
    }
}

fun waitForTextInView(text: String, @IdRes viewId: Int, timeout: Long) {
    val endTime = System.currentTimeMillis() + timeout
    var textInView = ""

    while (System.currentTimeMillis() < endTime) {
        textInView = getText(viewId)
        if (textInView == text) break
        sleepThread(500)
    }

    assertThat(textInView)
            .`as`("Expected text was not found in the View with ID '${resId(viewId)}' within $timeout millis")
            .isEqualTo(text)
}

fun waitForTextInView(@StringRes stringId: Int, @IdRes viewId: Int, timeout: Long) {
    waitForTextInView(resText(stringId), viewId, timeout)
}

fun waitForToast(matcher: Matcher<View>, timeout: Long = 20000) {
    val toastMatcher = object : TypeSafeMatcher<Root>() {
        override fun describeTo(description: Description) {
            description.appendText("is toast")
        }

        override fun matchesSafely(item: Root): Boolean {
            val type = item.windowLayoutParams.get().type
            @Suppress("DEPRECATION")
            if (type == WindowManager.LayoutParams.TYPE_TOAST) {
                val windowToken = item.decorView.windowToken
                val appToken = item.decorView.applicationWindowToken
                return windowToken === appToken
            }
            return false
        }
    }

    val endTime = System.currentTimeMillis() + timeout

    while (System.currentTimeMillis() < endTime) {
        try {
            onView(matcher).inRoot(toastMatcher).check(matches(isDisplayed()))
            return
        } catch (e: NoMatchingViewException) {
            sleepThread(500)
        } catch (e: AssertionFailedError) {
            sleepThread(500)
        }
    }

    throw TimeoutException("The Toast $matcher was not found within $timeout millis")
}

fun setChecked(matcher: Matcher<View?>, check: Boolean) {
    onView(matcher).perform(object : ViewAction {
        override fun getConstraints(): BaseMatcher<View> {
            return object : BaseMatcher<View>() {
                override fun matches(item: Any): Boolean {
                    return CoreMatchers.isA(Checkable::class.java).matches(item)
                }

                override fun describeMismatch(item: Any, mismatchDescription: Description) {}
                override fun describeTo(description: Description) {}
            }
        }

        override fun getDescription(): String? {
            return null
        }

        override fun perform(uiController: UiController, view: View) {
            val checkableView = view as Checkable

            if (check && !checkableView.isChecked) {
                view.performClick()
            } else if (!check && checkableView.isChecked) {
                view.performClick()
            }
        }
    })
}

fun setChecked(viewId: Int, check: Boolean) {
    setChecked(withId(viewId), check)
}

fun getText(matcher: Matcher<View>): String {
    var text = ""

    onView(matcher).perform(object : ViewAction {
        override fun getConstraints(): Matcher<View> {
            return isAssignableFrom(TextView::class.java)
        }

        override fun getDescription(): String {
            return String.format("getting text from the View: $matcher")
        }

        override fun perform(uiController: UiController, view: View) {
            text = (view as TextView).text.toString()
        }
    })

    return text
}

fun getText(@IdRes viewId: Int): String {
    return getText(withId(viewId))
}

fun getListViewItemsCount(@IdRes listViewId: Int): Int {
    var count = 0

    val matcher = object : TypeSafeMatcher<View>() {
        override fun describeTo(description: Description) {}

        override fun matchesSafely(item: View): Boolean {
            count = (item as ListView).count
            return true
        }
    }
    onView(withId(listViewId)).check(matches(matcher))

    return count
}

fun getListViewItemText(@IdRes listViewId: Int, position: Int): String {
    var text = ""

    onData(anything())
            .inAdapterView(withId(listViewId))
            .atPosition(position)
            .perform(object : ViewAction {
                override fun getConstraints(): Matcher<View> {
                    return isAssignableFrom(TextView::class.java)
                }

                override fun getDescription(): String {
                    return "getting text from a ListView item"
                }

                override fun perform(uiController: UiController, view: View) {
                    text = (view as TextView).text.toString()
                }
            })

    return text
}